### ISSUE TITLE
Download strategies should be able to be embedded in formulae 

### ISSUE THREAD

**Issue by MikeMcQuaid** [Homebrew/brew#5074]
https://github.com/Homebrew/brew/pull/5070 and https://github.com/Homebrew/brew/pull/5038 are PRs to add new download strategies that will not be used by Homebrew/homebrew-core. Instead of merging PRs like this we should require the addition of a refactoring which will allow download strategies to be created in formulae like in `radare2.rb`. 

After this we should deprecate and then remove all download strategies not used in Homebrew/homebrew-core in favour of allowing taps to maintain them themselves in formulae or in files `require`d by multiple formulae.

**Comment by MikeMcQuaid** [comment]
For example of a Requirement doing this see https://github.com/Homebrew/homebrew-core/blob/7d6810f30f3227ee6ee40b051f5bb6c6a288c366/Formula/radare2.rb#L1-L17

**Comment by colindean** [comment]
I think this is the way to go, too.

I'd like to see the download strategies refactored a good bit in the
process, too. I tried to add a download strategy about 2 years ago but
gave up because I couldn't really understand what was going on super
well, plus my use case may have been a little niche. Granted, that was
two years ago and I was doing it in haste!
Plugable download strategies would be great but a fear that I have is
continue all reimplantation of the same strategy if people can't
discover the ones they want someplace.
Would there be a way to have download strategies installable just
like formula?

**Comment by MikeMcQuaid** [comment]
> Would there be a way to have download strategies installable just like formula?

I don't think so. Basically pretty much all third-party download strategies are niche and I'm unconvinced the new proposed ones would be used by more than a handful for taps/organisations. If they duplicate work: fine.

**Comment by reitermarkus** [comment]
While refactoring the download strategies I had something like this in mind: 

```ruby
url "https://foo.bar/,
    using: lambda do |url, destination, **options|
      # custom download logic
    end
```

This way we don't expose the private internals which makes it easier to refactor.

**Comment by MikeMcQuaid** [comment]
@reitermarkus Yeh, something like that would be good 👍. It'd be nice if `using` could take a custom class, too, but that only requires a very limited number of methods to be defined.

**Comment by reitermarkus** [comment]
Yeah, you could pass anything that provides a `#call` method, like `lambda` and `proc` do.

**Comment by kralf** [comment]
Hi, as the author of #5308, I've initially considered doing something like this in the formula spec:
```
class MyFormula < Formula
  class MyDownloadStrategy < AbstractFileDownloadStrategy
    def initialize(url, name, version, **meta)
      super
       ...
     end

     def fetch
       ...
     end
  end

  ...

  bottle do
    root_url "https://foo.bar/", :using => MyDownloadStrategy
    ...
  end

  ...
end
```
This would work as expected as the download strategy can be specified by class name (given it inherits from `AbstractDownloadStrategy`). However, my reservation about this method is that a) it bloats the formula (rendering it unreadable) and b) it leads to duplication where the same strategy is required for multiple formulae of a (private) tap.

The lambda approach may suffer from similar problems if multiple resources in a formula adhere to the same download strategy. Furthermore, proper failure handling during an attempted download will often require more than just a view lines of code, so accommodating all this in a single block will again hurt style and readability.

For the above reasons, my preference would clearly be to maintain download strategy specs alongside with a tap's formulae and to then be able to `require` them in the formulae.

**Comment by MikeMcQuaid** [comment]
> For the above reasons, my preference would clearly be to maintain download strategy specs alongside with a tap's formulae and to then be able to `require` them in the formulae.

That should be possible in some form already by doing what you're doing in the formula spec and doing a `require_relative`. Let us know if that works for you or if there's an API issue here.
